lib/command/src/Obelisk/Command/Upgrade.hs

```diff
@@ -0,0 +1,186 @@
+import Obelisk.CliApp
+import Obelisk.Command.Utils
+
+import Obelisk.Command.Project (toImplDir)
+import Obelisk.Command.Project (findProjectObeliskCommand)
+import Obelisk.Command.Thunk (updateThunk)
+
+import Obelisk.Migration
+
+data MigrationGraph
+  = MigrationGraph_ObeliskUpgrade
+  | MigrationGraph_ObeliskHandoff
+  deriving (Eq, Show)
+
+graphName :: MigrationGraph -> Text
+graphName = \case
+  MigrationGraph_ObeliskHandoff -> "obelisk-handoff"
+  MigrationGraph_ObeliskUpgrade -> "obelisk-upgrade"
+
+fromGraphName :: Text -> MigrationGraph
+fromGraphName = \case
+  "obelisk-handoff" -> MigrationGraph_ObeliskHandoff
+  "obelisk-upgrade" -> MigrationGraph_ObeliskUpgrade
+  _ -> error "Invalid graph name specified"
+
+
+ensureCleanProject :: MonadObelisk m => FilePath -> m ()
+ensureCleanProject project = ensureCleanGitRepo project False "Cannot upgrade with uncommited changes"
+
+-- | Decide whether we (ambient ob) should handoff to project obelisk before performing upgrade
+decideHandOffToProjectOb :: MonadObelisk m => FilePath ->  m Bool
+decideHandOffToProjectOb project = do
+  ensureCleanProject project
+  updateThunk (toImplDir project) $ \projectOb -> do
+    getMigrationGraph projectOb MigrationGraph_ObeliskHandoff >>= \case
+      Nothing -> do
+        putLog Warning "Project obelisk is too old (has no migration graph); won't handoff"
+        return False
+      Just _ -> do
+        projectHash <- computeVertexHash projectOb MigrationGraph_ObeliskHandoff projectOb
+        (ambientGraph, ambientHash) <- getAmbientObInfo
+
+        case hasVertex projectHash ambientGraph of
+          False -> do
+            putLog Warning "Cannot find project ob in ambient ob's migration graph; handing off anyway"
+            return True
+          True -> case findPath projectHash ambientHash ambientGraph of
+            Nothing -> do
+              putLog Warning "No migration path between project and ambient ob; handing off anyway"
+              return True
+            Just ex -> do
+              actions <- sequence $ fmap (getEdge ambientGraph) ex
+              let dontHandoff = or $ flip fmap actions $ \case
+                    "True" -> True
+                    _ -> False
+              return $ not dontHandoff
+  where
+    getAmbientObInfo = do
+      ambientOb <- getAmbientOb
+      getMigrationGraph ambientOb MigrationGraph_ObeliskHandoff >>= \case
+        Nothing -> do
+          failWith "Ambient ob has no migration (this can't be possible)"
+        Just (m, _, ambientHash) -> do
+          unless (hasVertex ambientHash m) $
+            failWith "Ambient ob's hash is not in its own graph"
+          return (m, ambientHash)
+
+-- | Return the path to the current ('ambient') obelisk process Nix directory
+getAmbientOb :: MonadObelisk m => m FilePath
+getAmbientOb = takeDirectory . takeDirectory <$> liftIO getExecutablePath
+
+getEdge :: MonadObelisk m => Migration action -> (Hash, Hash) -> m action
```

Comment:
We should move all of the non-ob-specific migration graph stuff into a separate module where it can be reused.